{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2ab5f536",
   "metadata": {},
   "source": [
    "#### Project 1: Differential Expression Analysis (transcriptomics)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa95d9a5",
   "metadata": {},
   "source": [
    "The goal for this project is to investigate differences in gene expression between samples of healthy lung tissues and samples of idiopathic pulmonary fibrosis (IPF) lung tissue.\n",
    " - ([Data here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE150910))\n",
    " -  ([Paper here](https://pmc.ncbi.nlm.nih.gov/articles/PMC7667907/))\n",
    "\n",
    "**DISCLAIMER:**\n",
    "the purpose of this analysis is to be educational and simplistic, not to follow current best-practices for differential expression analysis. If you're doing this for real as part of a research project, you should probably use an R package like ([DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)),([edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html)) , or ([Limma-Voom](https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/thursday/DE.html)) and follow the standard analysis pipelines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98da86ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "%pip install scipy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "31a00512",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import csv\n",
    "import random\n",
    "import scipy\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import ks_2samp\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25c1a0f2",
   "metadata": {},
   "source": [
    "#### Step 1: Reading and processing data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc7a75ef",
   "metadata": {},
   "source": [
    "**Note:** This data is not included on this Github repo (because it's not my dataset) but can be downloaded for free on the ([Gene Expression Omnibus website here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE150910)). The specific file we need is called GSE150910_gene-level_count_file.csv.\n",
    "\n",
    "We'll store the gene names in a list called `genes`, and we'll store the expression measurements for each gene in a list of lists called `data`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "567247dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = []\n",
    "data = []\n",
    "with open(\"./data/GSE150910_gene-level_count_file.csv\") as csvfile:\n",
    "  reader = csv.reader(csvfile, delimiter=',')\n",
    "  for row in reader:\n",
    "    genes.append(row[0])\n",
    "    data.append(row[1:])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50fb757a",
   "metadata": {},
   "source": [
    "The first row was the sample labels, so we'll save that as its own list called `samples`, and then remove it from `data` and `genes`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "36ff1ba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = data[0]\n",
    "data = data[1:]\n",
    "genes = genes[1:]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99ba2511",
   "metadata": {},
   "source": [
    "Next, we'll cast everything to NumPy arrays, since those will be easier to work with than lists. And we also need to cast the elements of `data` to floats, since they were read in as strings by default."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "623847f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = np.array(data).astype(float)\n",
    "genes = np.array(genes)\n",
    "samples = np.array(samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53f4315f",
   "metadata": {},
   "source": [
    "Quick check to make sure the shapes of these arrays make sense..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "00c9bc8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18838,)\n",
      "(288,)\n",
      "(18838, 288)\n"
     ]
    }
   ],
   "source": [
    "print(genes.shape)\n",
    "print(samples.shape)\n",
    "print(data.shape)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5047a583",
   "metadata": {},
   "source": [
    "Quick check to make sure the content of these arrays looks of..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fa1395c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes:\n",
      "['TSPAN6' 'TNMD' 'DPM1' 'SCYL3' 'C1orf112']\n",
      "samples:\n",
      "['chp_26' 'chp_31' 'chp_34' 'chp_38' 'chp_1']\n",
      "data:\n",
      "[[1361.  993.  351. ...  465.  639.  944.]\n",
      " [   5.   13.    0. ...    0.    0.    0.]\n",
      " [1929. 2775. 1894. ...  860. 1387. 2150.]\n",
      " [ 176.  216.  208. ...   21.  146.  204.]\n",
      " [  93.  143.   97. ...   72.   69.  158.]]\n"
     ]
    }
   ],
   "source": [
    "print(\"genes:\")\n",
    "print(genes[0:5])\n",
    "print(\"samples:\")\n",
    "print(samples[0:5])\n",
    "print(\"data:\")\n",
    "print(data[0:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d37d4d6a",
   "metadata": {},
   "source": [
    "Okay, next we want to make an array with the experimental condition labels for each sample. We can do this with some pretty basic string manipulation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d17bcfe5",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = []\n",
    "\n",
    "for i in range(len(samples)):\n",
    "  tmp = samples[i].split(\"_\")\n",
    "  labels.append(tmp[0])\n",
    "    \n",
    "labels = np.array(labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32b0585e",
   "metadata": {},
   "source": [
    "Quick checks.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3c0e55a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chp' 'chp' 'chp' 'chp' 'chp']\n",
      "['chp' 'control' 'ipf']\n"
     ]
    }
   ],
   "source": [
    "print(labels[0:5])\n",
    "print(np.unique(labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b310d43d",
   "metadata": {},
   "source": [
    "Next, we'll get rid of the chronic hypersensitivity pneumonitis(CHP) samples so that we're left with only the control and idiopathic pulmonary fibrosis(IPF) samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d6fe4ac7",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data[:  ,labels != \"chp\"]\n",
    "samples = samples[labels != \"chp\"]\n",
    "labels = labels[labels != \"chp\"]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f01008de",
   "metadata": {},
   "source": [
    "quick sanity check again..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d1ae9eac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18838, 206)\n",
      "(206,)\n",
      "(206,)\n"
     ]
    }
   ],
   "source": [
    "print(data.shape)\n",
    "print(samples.shape)\n",
    "print(labels.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f63d522",
   "metadata": {},
   "source": [
    "The total number of counts in a sample is called **sequencing depth**. We want to control for the difference in sequencing depth between samples, in case the experiment happened to be more sensitive in some samples and less in others.\n",
    "To do this, we will apply a counts per million(CPM) normalization. For each sample, we'll divide the counts for each gene by the total number of counts in that sample, then multiply by a million."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "46995a03",
   "metadata": {},
   "outputs": [],
   "source": [
    "for j in range(data.shape[1]):\n",
    "  columnSum = sum(data[:, j])\n",
    "  \n",
    "  data[:,j] = data[:,j] / columnSum * 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5bc4e0a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[2.62026354e+01 3.98260770e+01 3.24569396e+01 ... 3.31761577e+01\n",
      "  3.12351859e+01 3.07286955e+01]\n",
      " [0.00000000e+00 0.00000000e+00 0.00000000e+00 ... 1.82286581e-01\n",
      "  0.00000000e+00 0.00000000e+00]\n",
      " [7.37006988e+01 8.12189622e+01 5.73764856e+01 ... 6.33142058e+01\n",
      "  6.77984395e+01 6.99859061e+01]\n",
      " ...\n",
      " [1.23451757e-01 2.91499191e-01 1.92280448e-01 ... 3.03810968e-02\n",
      "  1.46644065e-01 1.62757921e-01]\n",
      " [1.82091341e+00 2.76924232e+00 1.42287531e+00 ... 1.79248471e+00\n",
      "  1.80861014e+00 2.05074981e+00]\n",
      " [0.00000000e+00 0.00000000e+00 0.00000000e+00 ... 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00]]\n"
     ]
    }
   ],
   "source": [
    "print(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b4360d6",
   "metadata": {},
   "source": [
    "This is one common way of normalization gene expression data. Some other methods also control for gene length (which we don't have for this dataset) in addition to sequencing depth. \n",
    "Next we'll apply a filter to eliminate genes that had very low expression in both conditions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a35acd64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11097, 206)\n",
      "11097\n"
     ]
    }
   ],
   "source": [
    "mean_CPM_Control = data[:,labels==\"control\"].mean(axis=1)\n",
    "mean_CPM_ipf = data[:, labels == \"ipf\"].mean(axis=1)\n",
    "\n",
    "toKeep = (mean_CPM_ipf >= 5) | (mean_CPM_Control >= 5)\n",
    "\n",
    "data=data[toKeep, :]\n",
    "genes = genes[toKeep]\n",
    "\n",
    "print(data.shape)\n",
    "print(len(genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faa3e4c8",
   "metadata": {},
   "source": [
    "**NOTE:** in a real-life research project, you might want to apply additional quality checks and normalizations at this step -- batch correction, correction based on demographic info about subjects (sex, race, age, health status, etc), identifying outliers and thinking about if you should remove them or not. For the purposes of this class, we'll skip these steps, but in a real project this would be the time to do them."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eddf986",
   "metadata": {},
   "source": [
    "#### Step 2: Differential expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fb0c295",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (GPU_Env)",
   "language": "python",
   "name": "gpu_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
