{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2ab5f536",
   "metadata": {},
   "source": [
    "#### Project 1: Differential Expression Analysis (transcriptomics)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa95d9a5",
   "metadata": {},
   "source": [
    "The goal for this project is to investigate differences in gene expression between samples of healthy lung tissues and samples of idiopathic pulmonary fibrosis (IPF) lung tissue.\n",
    " - ([Data here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE150910))\n",
    " -  ([Paper here](https://pmc.ncbi.nlm.nih.gov/articles/PMC7667907/))\n",
    "\n",
    "**DISCLAIMER:**\n",
    "the purpose of this analysis is to be educational and simplistic, not to follow current best-practices for differential expression analysis. If you're doing this for real as part of a research project, you should probably use an R package like ([DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)),([edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html)) , or ([Limma-Voom](https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/thursday/DE.html)) and follow the standard analysis pipelines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98da86ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "%pip install scipy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "31a00512",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import csv\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import ks_2samp\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25c1a0f2",
   "metadata": {},
   "source": [
    "#### Step 1: Reading and processing data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc7a75ef",
   "metadata": {},
   "source": [
    "**Note:** This data is not included on this Github repo (because it's not my dataset) but can be downloaded for free on the ([Gene Expression Omnibus website here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE150910)). The specific file we need is called GSE150910_gene-level_count_file.csv.\n",
    "\n",
    "We'll store the gene names in a list called `genes`, and we'll store the expression measurements for each gene in a list of lists called `data`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "567247dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = []\n",
    "data = []\n",
    "with open(\"./data/GSE150910_gene-level_count_file.csv\") as csvfile:\n",
    "  reader = csv.reader(csvfile, delimiter=',')\n",
    "  for row in reader:\n",
    "    genes.append(row[0])\n",
    "    data.append(row[1:])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50fb757a",
   "metadata": {},
   "source": [
    "The first row was the sample labels, so we'll save that as its own list called `samples`, and then remove it from `data` and `genes`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "36ff1ba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = data[0]\n",
    "data = data[1:]\n",
    "genes = genes[1:]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99ba2511",
   "metadata": {},
   "source": [
    "Next, we'll cast everything to NumPy arrays, since those will be easier to work with than lists. And we also need to cast the elements of `data` to floats, since they were read in as strings by default."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "623847f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = np.array(data).astype(float)\n",
    "genes = np.array(genes)\n",
    "samples = np.array(samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53f4315f",
   "metadata": {},
   "source": [
    "Quick check to make sure the shapes of these arrays make sense..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "00c9bc8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18838,)\n",
      "(288,)\n",
      "(18838, 288)\n"
     ]
    }
   ],
   "source": [
    "print(genes.shape)\n",
    "print(samples.shape)\n",
    "print(data.shape)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5047a583",
   "metadata": {},
   "source": [
    "Quick check to make sure the content of these arrays looks of..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fa1395c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes:\n",
      "['TSPAN6' 'TNMD' 'DPM1' 'SCYL3' 'C1orf112']\n",
      "samples:\n",
      "['chp_26' 'chp_31' 'chp_34' 'chp_38' 'chp_1']\n",
      "data:\n",
      "[[1361.  993.  351. ...  465.  639.  944.]\n",
      " [   5.   13.    0. ...    0.    0.    0.]\n",
      " [1929. 2775. 1894. ...  860. 1387. 2150.]\n",
      " [ 176.  216.  208. ...   21.  146.  204.]\n",
      " [  93.  143.   97. ...   72.   69.  158.]]\n"
     ]
    }
   ],
   "source": [
    "print(\"genes:\")\n",
    "print(genes[0:5])\n",
    "print(\"samples:\")\n",
    "print(samples[0:5])\n",
    "print(\"data:\")\n",
    "print(data[0:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d37d4d6a",
   "metadata": {},
   "source": [
    "Okay, next we want to make an array with the experimental condition labels for each sample. We can do this with some pretty basic string manipulation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d17bcfe5",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = []\n",
    "\n",
    "for i in range(len(samples)):\n",
    "  tmp = samples[i].split(\"_\")\n",
    "  labels.append(tmp[0])\n",
    "    \n",
    "labels = np.array(labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32b0585e",
   "metadata": {},
   "source": [
    "Quick checks.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3c0e55a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chp' 'chp' 'chp' 'chp' 'chp']\n",
      "['chp' 'control' 'ipf']\n"
     ]
    }
   ],
   "source": [
    "print(labels[0:5])\n",
    "print(np.unique(labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b310d43d",
   "metadata": {},
   "source": [
    "Next, we'll get rid of the chronic hypersensitivity pneumonitis(CHP) samples so that we're left with only the control and idiopathic pulmonary fibrosis(IPF) samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d6fe4ac7",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data[:  ,labels != \"chp\"]\n",
    "samples = samples[labels != \"chp\"]\n",
    "labels = labels[labels != \"chp\"]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f01008de",
   "metadata": {},
   "source": [
    "quick sanity check again..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d1ae9eac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18838, 206)\n",
      "(206,)\n",
      "(206,)\n"
     ]
    }
   ],
   "source": [
    "print(data.shape)\n",
    "print(samples.shape)\n",
    "print(labels.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f63d522",
   "metadata": {},
   "source": [
    "The total number of counts in a sample is called **sequencing depth**. We want to control for the difference in sequencing depth between samples, in case the experiment happened to be more sensitive in some samples and less in others.\n",
    "To do this, we will apply a counts per million(CPM) normalization. For each sample, we'll divide the counts for each gene by the total number of counts in that sample, then multiply by a million.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1532ad8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "206\n"
     ]
    }
   ],
   "source": [
    "print(data.shape[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "46995a03",
   "metadata": {},
   "outputs": [],
   "source": [
    "for j in range(data.shape[1]):\n",
    "  columnSum = sum(data[:, j])\n",
    "  \n",
    "  data[:,j] = data[:,j] / columnSum * 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5bc4e0a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[2.62026354e+01 3.98260770e+01 3.24569396e+01 ... 3.31761577e+01\n",
      "  3.12351859e+01 3.07286955e+01]\n",
      " [0.00000000e+00 0.00000000e+00 0.00000000e+00 ... 1.82286581e-01\n",
      "  0.00000000e+00 0.00000000e+00]\n",
      " [7.37006988e+01 8.12189622e+01 5.73764856e+01 ... 6.33142058e+01\n",
      "  6.77984395e+01 6.99859061e+01]\n",
      " ...\n",
      " [1.23451757e-01 2.91499191e-01 1.92280448e-01 ... 3.03810968e-02\n",
      "  1.46644065e-01 1.62757921e-01]\n",
      " [1.82091341e+00 2.76924232e+00 1.42287531e+00 ... 1.79248471e+00\n",
      "  1.80861014e+00 2.05074981e+00]\n",
      " [0.00000000e+00 0.00000000e+00 0.00000000e+00 ... 0.00000000e+00\n",
      "  0.00000000e+00 0.00000000e+00]]\n"
     ]
    }
   ],
   "source": [
    "print(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b4360d6",
   "metadata": {},
   "source": [
    "This is one common way of normalization gene expression data. Some other methods also control for gene length (which we don't have for this dataset) in addition to sequencing depth. \n",
    "Next we'll apply a filter to eliminate genes that had very low expression in both conditions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a35acd64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11097, 206)\n",
      "11097\n"
     ]
    }
   ],
   "source": [
    "mean_CPM_Control = data[:,labels==\"control\"].mean(axis=1)\n",
    "mean_CPM_ipf = data[:, labels == \"ipf\"].mean(axis=1)\n",
    "\n",
    "toKeep = (mean_CPM_ipf >= 5) | (mean_CPM_Control >= 5)\n",
    "\n",
    "data=data[toKeep, :]\n",
    "genes = genes[toKeep]\n",
    "\n",
    "print(data.shape)\n",
    "print(len(genes))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faa3e4c8",
   "metadata": {},
   "source": [
    "**NOTE:** in a real-life research project, you might want to apply additional quality checks and normalizations at this step -- batch correction, correction based on demographic info about subjects (sex, race, age, health status, etc), identifying outliers and thinking about if you should remove them or not. For the purposes of this class, we'll skip these steps, but in a real project this would be the time to do them."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eddf986",
   "metadata": {},
   "source": [
    "#### Step 2: Differential expression analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a36c4422",
   "metadata": {},
   "source": [
    "Now, we'll check each gene to see if it's expression is different in the control and IPF conditions. The standard R liraries for differential expression analysis use complex statistical models to check this, but for our educational purposes, we'll just use a ([Kolmogorv-Smirnov test](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test)). The KS test is non-paremetric, meaning that it doesn't make any assumptions about the distribution being compared.\n",
    "\n",
    "Note that using the KS test here is unrealistic and it's not typically used in real analysis pipelines,not because it's incorrect, but because it's underpowered to detect true differences in expression compared to the more clever statistical models. But the goal for this class isn't to follow a realistic analysis pipeline, it's to explore this dataset in a way that every step of the analysis makes sense to us, as an educational exercise.\n",
    "\n",
    "Ok, let's try to actually do this differential expression check now. We're eventually going to iterate over the dataset and check every gene, but let's just start with one gene (the first, at index 0) for now so we can figure out what we need to do.\n",
    "\n",
    "To start, we'll split the distribution into two separate arrays: one with the gene expression for control samples, and the other with gene expression for the IPF samples:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1fb0c295",
   "metadata": {},
   "outputs": [],
   "source": [
    "#starting with just the first gene. We'll eventually do this for every gene\n",
    "control = data[0, labels == \"control\"]\n",
    "ipf = data[0, labels == \"ipf\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24f3bfb1",
   "metadata": {},
   "source": [
    "Quick shape check..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d1e3ac1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(103,)\n",
      "(103,)\n"
     ]
    }
   ],
   "source": [
    "print(control.shape)\n",
    "print(ipf.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2b26c86",
   "metadata": {},
   "source": [
    "Next, we'll actually perform thr KS-test, using `ks_2samp` function that we imported from `scipy.stats` at the top, and remembering to add our `epsilon` value to every element and apply the log2 transformation. This gives us two values: a test statistic and a p-value: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fd904be0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import ks_2samp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ca392ef3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ks_statistic: 0.4368932038834951\n",
      "p_value: 3.264129457102852e-09\n"
     ]
    }
   ],
   "source": [
    "from scipy.stats import ks_2samp\n",
    "ksStatistics,pValue = ks_2samp(control, ipf)\n",
    "\n",
    "print(f\"ks_statistic: {ksStatistics}\")\n",
    "print(f\"p_value: {pValue}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "934e9fd7",
   "metadata": {},
   "source": [
    "Again, this is just a test of whether the distriubtions are different, and doesn't tell us whether or not there's a difference in means (which is what we're really interested in). So besides getting the p-value for the statistical test, we also want to calculate something called the $\\log_2$(fold change) between the mean expression level in each condition:\n",
    "\n",
    "$$\n",
    "\\log_2 \\left( \\frac{\\text{Expression Mean (IPF)}}{\\text{Expression Mean (Control)}} \\right)\n",
    "$$\n",
    "\n",
    "The inside fraction $\\frac{\\text{Expression Mean (IPF)}}{\\text{Expression Mean (Control)}}$ is the fold change. This tells us how different the mean expression of the IPF samples is compared to the control samples. The reason we're using $\\log_2$ is to make this measurement symetrical around 0. So if a gene's mean expression in IPF samples is *twice* the mean of the control samples, then $\\log_2 \\left( \\frac{\\text{Expression Mean (IPF)}}{\\text{Expression Mean (Control)}} \\right) = 1$. If it's *half* the mean of the control samples, then $\\log_2 \\left( \\frac{\\text{Expression Mean (IPF)}}{\\text{Expression Mean (Control)}} \\right) = -1$.\n",
    "\n",
    "One more thing before we actually calculate this -- if either of the means is 0, that will cause a problem for us. We can't divide by 0, and we also can't do $\\log_2(0)$. We'll deal with this by defining a small constant called `epsilon` (often 1, 0.1, or 0.01) and adding it to both means before doing the calculation.\n",
    "\n",
    "Here's how this calculation is done, again using the first gene (index 0) as an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a7bde5f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7894289821093974\n"
     ]
    }
   ],
   "source": [
    "epsilon = 1 #small constant to avoid division by zero\n",
    "\n",
    "controlMean = np.mean(control)\n",
    "ipfMean = np.mean(ipf)\n",
    "\n",
    "# calculate log2(fold change), adding epsilon to avoid division by 0\n",
    "FC = (ipfMean+epsilon) / (controlMean+epsilon)\n",
    "log2_FC = np.log2(FC)\n",
    "print(log2_FC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46a90a37",
   "metadata": {},
   "source": [
    "ok, that's basically it! Now we just need to go through and do this to every gene, and save `p_value` and `log2_FC` for each one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c7fb6883",
   "metadata": {},
   "outputs": [],
   "source": [
    "#getting p-value and log2FC, now for every gene\n",
    "\n",
    "p_values = []\n",
    "log2_FCs = []\n",
    "\n",
    "epsilon = 1\n",
    "\n",
    "for i in range(data.shape[0]):\n",
    "  control = data[i, labels == \"control\"]\n",
    "  ipf = data[i, labels == \"ipf\"]\n",
    "  \n",
    "  # perform ks-test\n",
    "  ksStatistics, pValue = ks_2samp(control,ipf)\n",
    "  \n",
    "  # save p_value for this gene\n",
    "  p_values.append(pValue)\n",
    "  \n",
    "  # calculate means for each condition\n",
    "  controlMean = np.mean(control)\n",
    "  ipfMean = np.mean(ipf)\n",
    "  \n",
    "  # calculate log2(fold change), adding epsilon to avoid division by 0\n",
    "  FC = (ipfMean+epsilon) / (controlMean+epsilon)\n",
    "  log2_FC = np.log2(FC)\n",
    "  \n",
    "  # save log2_Fc for this gene\n",
    "  log2_FCs.append(log2_FC)\n",
    "  \n",
    "p_values = np.array(p_values)\n",
    "log2_FCs = np.array(log2_FCs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "18cdc47c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[3.26412946e-09 1.79707436e-02 6.02011471e-01 ... 2.28643986e-05\n",
      " 2.95016347e-04 1.14321993e-05]\n"
     ]
    }
   ],
   "source": [
    "print(p_values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa768d0d",
   "metadata": {},
   "source": [
    "Next we'll do a [Bonferroni correction](https://en.wikipedia.org/wiki/Bonferroni_correction), multiplying each p-value by the number of tests we did. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "49087205",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_values_bonf = p_values * len(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "da805c2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[3.62220446e-05 1.99421342e+02 6.68052129e+03 ... 2.53726232e-01\n",
      " 3.27379640e+00 1.26863116e-01]\n"
     ]
    }
   ],
   "source": [
    "print(p_values_bonf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d432b611",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "(**NOTE**: in practice, people typically use another multiple-test correction called the [ Benjamini-Hochberg correction](https://www.youtube.com/watch?v=K8LQSvtjcEo), which is more lenient. But for educational purposes, we'll use the Bonferroni correction, since it's easier to understand on an inuitive level.)\n",
    "\n",
    "Now we can use our `p_values_bonf` and `log2_FCs` arrays to select a list of differentially expressed genes (DEGs) between the two conditions. The threshold choice is somewhat arbitrary here. I'm going to use `p_values_bonf <= 0.05` for statistical significance, and `np.abs(log2_FCs) >= 2` to select genes with a very large effect size (the IPF mean must be greater than 4x or less than 1/4 the control mean).\n",
    "\n",
    "This `np.abs(log2_FCs) >= 2` cutoff is probably too strict, and in a real-life scenario it would probably make more sense to use `np.abs(log2_FCs) >= 1`. But for this dataset,` np.abs(log2_FCs) >= 1` yields quite a lot of DEGs, so for the purposes of this demo I'm trying to narrow down the list to make the next plot we're going to do easier.\n",
    "\n",
    "(**NOTE**: sometimes the combination of p-values and log2_FC like this is called biological significance, since it requires statistical significance (determined by the p-value) as well as biological relevance (determined by the log2_FC, which measures the effect size).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "55fc136b",
   "metadata": {},
   "outputs": [],
   "source": [
    "to_keep = (p_values_bonf <= 0.05) & (np.abs(log2_FCs) >= 2)\n",
    "\n",
    "sig_genes = genes[to_keep]\n",
    "sig_log2_FCs = log2_FCs[to_keep]\n",
    "sig_p_values_bonf = p_values_bonf[to_keep]\n",
    "sig_data = data[to_keep, :]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de6cae65",
   "metadata": {},
   "source": [
    "Some quick checks to see how many DEGs this yielded:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f595c5b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of DEGs: 79\n",
      "Total number of genes: 11097\n",
      "DEG percentage: 0.7119041182301523\n",
      "\n",
      "Number of upregulated DEGs: 65\n",
      "Number of downregulated DEGs: 14\n"
     ]
    }
   ],
   "source": [
    "print(f\"Number of DEGs: {len(sig_genes)}\")\n",
    "print(f\"Total number of genes: {len(genes)}\")\n",
    "print(f\"DEG percentage: {len(sig_genes) / len(genes) * 100}\")\n",
    "print()\n",
    "print(f\"Number of upregulated DEGs: {np.sum(sig_log2_FCs>0)}\")\n",
    "print(f\"Number of downregulated DEGs: {np.sum(sig_log2_FCs<0)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01d68b78",
   "metadata": {},
   "source": [
    "if we want, we can also print out the list of DEGs to take a look. This might also be a good time to check [Enrichr](https://maayanlab.cloud/Enrichr) to see if there's anything interesting about them."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8ceeed1",
   "metadata": {},
   "source": [
    "*Upregulated Genes*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8c81c228",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PROM1\n",
      "CYP24A1\n",
      "CP\n",
      "CDH3\n",
      "COL17A1\n",
      "ATP12A\n",
      "FAT2\n",
      "SLC4A11\n",
      "DERL3\n",
      "VSIG1\n",
      "CD79A\n",
      "COMP\n",
      "COL1A1\n",
      "COL7A1\n",
      "FHL2\n",
      "TSPAN1\n",
      "MUC5B\n",
      "SPP1\n",
      "IL13RA2\n",
      "COL10A1\n",
      "BPIFB1\n",
      "PDLIM4\n",
      "TNS4\n",
      "FCRLA\n",
      "JCHAIN\n",
      "POSTN\n",
      "ERN2\n",
      "TMPRSS4\n",
      "STRA6\n",
      "GPR87\n",
      "FCRL5\n",
      "MUC4\n",
      "CXCL14\n",
      "FAM83A\n",
      "PLEKHS1\n",
      "SYT8\n",
      "THY1\n",
      "CXCL13\n",
      "MS4A1\n",
      "SCGB3A1\n",
      "CTHRC1\n",
      "CLDN2\n",
      "GREM1\n",
      "UGT1A6\n",
      "COL3A1\n",
      "SPRR1A\n",
      "MZB1\n",
      "KRT15\n",
      "LGALS7B\n",
      "B3GNT3\n",
      "MUC16\n",
      "FDCSP\n",
      "ALDH1A3\n",
      "COL14A1\n",
      "PLA2G2A\n",
      "S100A2\n",
      "BPIFA1\n",
      "IGFL2\n",
      "LGALS7\n",
      "SERPINB4\n",
      "DIO2\n",
      "IGLL5\n",
      "AC136428.1\n",
      "MSMB\n",
      "PRSS2\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(sig_genes)):\n",
    "  if sig_log2_FCs[i] > 0:\n",
    "    print(sig_genes[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51234c61",
   "metadata": {},
   "source": [
    "*Downregulated Genes*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8b40ef03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PRX\n",
      "SLC6A4\n",
      "MYRF\n",
      "FCN3\n",
      "GGTLC1\n",
      "ITLN2\n",
      "DEFA4\n",
      "BTNL9\n",
      "MS4A15\n",
      "CA4\n",
      "RTKN2\n",
      "AGER\n",
      "DEFA1\n",
      "DEFA1B\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(sig_genes)):\n",
    "  if sig_log2_FCs[i] < 0:\n",
    "    print(sig_genes[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6332f6b",
   "metadata": {},
   "source": [
    "Finally, we'll save these results to CSV files -- one for the list of downregulated DEGs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "fd0a30e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import csv\n",
    "\n",
    "# Create output directory if it does not exist\n",
    "os.makedirs(\"./output\", exist_ok=True)\n",
    "# saving upregulated genes\n",
    "outputFile = \"./output/upregulated_DEGs.csv\"\n",
    "with open(outputFile, mode = 'w', newline='') as file:\n",
    "  writer = csv.writer(file)\n",
    "  writer.writerow([\"gene\", \"p_value\", \"log2_FC\"])\n",
    "  \n",
    "  for i in range(len(sig_genes)):\n",
    "    if sig_log2_FCs[i] > 0:\n",
    "      writer.writerow([sig_genes[i], sig_p_values_bonf[i], sig_log2_FCs[i]])\n",
    "      \n",
    "# saving downregulated genes\n",
    "outputFile = \"./output/downregulated_DEGs.csv\"\n",
    "with open(outputFile, mode = 'w', newline='') as file:\n",
    "  writer = csv.writer(file)\n",
    "  writer.writerow([\"gene\", \"p_value\", \"log2_FC\"])\n",
    "  \n",
    "  for i in range(len(sig_genes)):\n",
    "    if sig_log2_FCs[i] < 0:\n",
    "      writer.writerow([sig_genes[i], sig_p_values_bonf[i], sig_log2_FCs[i]])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a45772b9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (TF-GPU)",
   "language": "python",
   "name": "tf-gpu"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
